### ISSUE TITLE
Hallucination/ code mangeling 

### ISSUE THREAD

**Issue by Kvadratni** [block/goose#60]
Sometimes goose experiences a hallucination as most LLM's however due to it's agent nature it can destroy code it was working on. 

![image](https://github.com/user-attachments/assets/178e4f58-8ae8-4af2-9be4-df087cf03abb)

Several fixes for this may be:
1.  Run the returned code agains a compiler before making the change
2.  Run a validator agent loop

**Comment by michaelneale** [comment]
@baxen this was a bit like I remember you discussing, the benefit of a fast local look post goose editing a file - if it is syntactically nonsense in some fast detectable way, can give that maybe as fast feedback to goose to try again.

One thing to try is tree sitter: 

https://github.com/grantjenks/py-tree-sitter-languages

and just see how it parses quickly per edit

**Comment by TimerErTim** [comment]
For brainstorming purposes, I think another viable solution would be integration with git. A goose session starts by branching off from the current branch and then commits changes for every individual step it does.

This way you can easily revise to previously working state and also keep track of the individual changes made. This way you could f.e. manually combine the end result with some interesting step that was lost in the end product.

**Comment by michaelneale** [comment]
@TimerErTim yeah - I have informally done that with goose in the past (.goosehints for example) - so yeah can totally work. Could be done with an MCP extension (ie specific clear function calls to start/end an edit batch - which it must always call). And it could also then deliberately not work if there are unstaged changes (so you don't lose any work?)
